CCDB/src/CcdbApi.cxx

```diff
@@ -424,14 +535,195 @@ TObject* CcdbApi::retrieveFromTFile(std::string path, std::map<std::string, std:
+{
+  if (!boost::filesystem::exists(filename)) {
+    LOG(INFO) << "Local snapshot " << filename << " not found \n";
+    return nullptr;
+  }
+  TFile f(filename.c_str(), "READ");
+  auto* object = f.GetObjectChecked(objname.c_str(), tcl);
+  // do some other stuff like copying if necessary
+
+  return object;
+}
+
+void* CcdbApi::retrieveFromTFile(std::type_info const& tinfo, std::string const& path,
+                                 std::map<std::string, std::string> const& metadata, long timestamp) const
+{
+  // We need the TClass for this type; will verify if dictionary exists
+  auto tcl = TClass::GetClass(tinfo);
+  if (!tcl) {
+    std::cerr << "Could not retrieve ROOT dictionary for type " << tinfo.name() << " aborting to read from CCDB";
+    return nullptr;
+  }
+  string objectName = string(tcl->GetName());
+  utils::trim(objectName);
+
+  // Note : based on https://curl.haxx.se/libcurl/c/getinmemory.html
+  // Thus it does not comply to our coding guidelines as it is a copy paste.
+
+  string fullUrl = getFullUrlForRetrieval(path, metadata, timestamp);
+
+  // if we are in snapshot mode we can simply open the file; extract the object and return
+  if (mInSnapshotMode) {
+    return extractFromLocalFile(fullUrl, objectName, tcl);
+  }
+
+  // Prepare CURL
+  CURL* curl_handle;
+  CURLcode res;
+  struct MemoryStruct chunk {
+    (char*)malloc(1) /*memory*/, 0 /*size*/
+  };
+
+  /* init the curl session */
+  curl_handle = curl_easy_init();
+
+  /* specify URL to get */
+  curl_easy_setopt(curl_handle, CURLOPT_URL, fullUrl.c_str());
+
+  /* send all data to this function  */
+  curl_easy_setopt(curl_handle, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
+
+  /* we pass our 'chunk' struct to the callback function */
+  curl_easy_setopt(curl_handle, CURLOPT_WRITEDATA, (void*)&chunk);
+
+  /* some servers don't like requests that are made without a user-agent
+         field, so we provide one */
+  curl_easy_setopt(curl_handle, CURLOPT_USERAGENT, "libcurl-agent/1.0");
+
+  /* if redirected , we tell libcurl to follow redirection */
+  curl_easy_setopt(curl_handle, CURLOPT_FOLLOWLOCATION, 1L);
+
+  /* get it! */
+  res = curl_easy_perform(curl_handle);
+
+  void* result = nullptr;
+  if (res == CURLE_OK) {
+    long response_code;
+    res = curl_easy_getinfo(curl_handle, CURLINFO_RESPONSE_CODE, &response_code);
+    if ((res == CURLE_OK) && (response_code != 404)) {
+      Int_t previousErrorLevel = gErrorIgnoreLevel;
+      gErrorIgnoreLevel = kFatal;
+      TMemFile memFile("name", chunk.memory, chunk.size, "READ");
+      gErrorIgnoreLevel = previousErrorLevel;
+      if (!memFile.IsZombie()) {
+        void* object = memFile.GetObjectChecked(objectName.c_str(), tcl);
+        if (object != nullptr) {
+          result = object;
+          //              // Copy object in memory and make sure we detach it from the file
```

Comment:
Have you tried whether it works with a TH1 ? or worse a TTree ?

Reply:
will do now

Reply:
This now works. Made object extraction and handling of particular cases a standalone function which is used in multiple places.